torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+from torch.optim.optimizer import register_optimizer_step_post_hook
+from torch.utils._python_dispatch import TorchDispatchMode
+from torch.utils._pytree import tree_flatten, tree_map_only
+from torch.utils.hooks import RemovableHandle
+from torch.utils.module_tracker import ModuleTracker
+from torch.utils.weak import WeakIdKeyDictionary
+
+
+_PYTORCH_MIN_ALLOCATE = 2**9
+_TOTAL = "Tracker Total"
+
+__all__ = ["MemTracker"]
+
+
+class RefType(str, Enum):
+    pass
+
+
+class MemRefType(RefType):
+    parameter = "Parameter"
+    buffer = "Buffer"
+    gradient = "Gradient"
+    activation = "Activation"
+    temp = "Temp"
+    optstate = "Optstate"
+    other = "Other"
+
+
+class ModState(str, Enum):
+    pre_fw = "Pre-Forward"
+    post_fw = "Post-Forward"
+    pre_bw = "Pre-Backward"
+    post_bw = "Post-Backward"
+    pre_fw_ac = "Pre-Forward (AC)"
+    post_fw_ac = "Post-Forward (AC)"
+    peak_fw = "Peak Forward"
+    peak_bw = "Peak Backward"
+
+
+@dataclass
+class ModMemStats:
+    """
+    A class to store the memory statistics of a module.
+    Args:
+        mod_fqn (str): The fully qualified name of the module.
+    Attributes:
+        mod_fqn (str): The fully qualified name of the module.
+        parameter_mem (int): The memory usage of the parameters of the module.
+        buffer_mem (int): The memory usage of the buffers of the module.
+        input_mem (int): The memory usage of the inputs to the module.
+        output_mem (int): The memory usage of the outputs from the module.
+        snapshots (Dict[ModState, Dict[torch.device, Dict[str, int]]]): A dictionary of memory snapshots
+        of the module at different states [pre_fw, post_fw, pre_bw, post_bw, peak_fw, peak_bw].
+    Note:
+        The memory snapshot is stored as a dictionary - Dict[torch.device, Dict[str, int]], where each key is a device,
+         and each value is another dictionary with keys as memory reference types (e.g., parameters, buffers,
+         activations, gradients, optimizer states) and values as the memory consumed in bytes.
+    """
+
+    def __init__(self, mod_fqn: str):
+        self.mod_fqn = mod_fqn
+        self.parameter_mem: int
+        self.buffer_mem: int
+        self.input_mem: int
+        self.output_mem: int
+        self.local_max: Dict[torch.device, int]
+        self.snapshots: Dict[ModState, Dict[torch.device, Dict[str, int]]] = {}
+
+
+@dataclass
+class WeakRefInfo:
+    def __init__(
+        self, size: int, element_size: int, device: torch.device, reftype: RefType
+    ) -> None:
+        self.size = size
+        self.element_size = element_size
+        self.reftype = reftype
+        self.device = device
+        self.mem_consumed = self._calculate_mem_consumed()
+
+    def _calculate_mem_consumed(self) -> int:
+        mem = self.size * self.element_size
+        if self.device.type == "cuda":
+            return math.ceil((mem) / _PYTORCH_MIN_ALLOCATE) * _PYTORCH_MIN_ALLOCATE
```

Comment:
I tested on cpu, it doesn't require a min-allocate as it allocates to byte level. Haven't used torch builds on other devices. Do you know where can I get this information?

Reply:
cc @jeffdaily is there a central place this information might exist? Or should we read the allocator code for each device?

Reply:
What is the point of this code? 
 _PYTORCH_MIN_ALLOCATE is hardcoded above to 2**9 (512), only used for this "cuda" [rocm] case.  Is this because CUDA aligns its pointers to 512 bytes (cudaDeviceProp::textureAlignment).  Or is it because c10/cuda/CUDACachingAllocator.cpp kMinBlockSize = 512?  What if the cuda caching allocator is disabled, does this value change?

Reply:
@jeffdaily this is a great question. The _PYTORCH_MIN_ALLOCATE is for c10/cuda/CUDACachingAllocator.cpp kMinBlockSize = 512. I haven't tried by disabling the caching allocator. If you have some pointers on that we can generalize this.